sdks/python/apache_beam/transforms/stats.py

```diff
@@ -234,3 +239,419 @@ def extract_output(accumulator):
+  in the natural order.
+
+  To evaluate the quantiles, we use the "New Algorithm" described here:
+
+  [MRL98] Manku, Rajagopalan & Lindsay, "Approximate Medians and other
+  Quantiles in One Pass and with Limited Memory", Proc. 1998 ACM SIGMOD,
+  Vol 27, No 2, p 426-435, June 1998.
+  http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.6.6513&rep=rep1
+  &type=pdf
+
+  The default error bound is (1 / N), though in practice the accuracy
+  tends to be much better.
+
+  Args:
+    num_quantiles: Number of quantiles to produce. It is the size of the final
+      output list, including the mininum and maximum value items.
+    buffer_size: The size of the buffers, corresponding to k in the referenced
+      paper.
+    num_buffers: The number of buffers, corresponding to b in the referenced
+      paper.
+    compare: (optional) Comparator function which is an implementation
+        of "a < b" taking two arguments (a and b). Which is later
+        converted to key function as Python 3 does not support `cmp`.
+    key: (optional) Key is a mapping of elements to a comparable key, similar
+      to the key argument of Python's sorting methods.
+    reverse: (optional) whether to order things smallest to largest, rather
+        than largest to smallest
+  """
+
+  # For alternating between biasing up and down in the above even weight
+  # collapse operation.
+  _offset_jitter = 0
+
+  # The cost (in time and space) to compute quantiles to a given accuracy is a
+  # function of the total number of elements in the data set. If an estimate is
+  # not known or specified, we use this as an upper bound. If this is too low,
+  # errors may exceed the requested tolerance; if too high, efficiency may be
+  # non-optimal. The impact is logarithmic with respect to this value, so this
+  # default should be fine for most uses.
+  _MAX_NUM_ELEMENTS = 1e9
+  _qs = None  # Refers to the _QuantileState
+
+  # TODO(mszb): For Python 3, remove compare and only keep key.
+  def __init__(self, num_quantiles, buffer_size, num_buffers, compare=None,
+               key=None, reverse=False):
+    if compare:
+      warnings.warn('Compare is not available in Python 3, use key instead.')
+
+    if key:
+      self._comparator = lambda a, b: (key(a) < key(b)) - (key(a) > key(b)) \
+        if reverse else (key(a) > key(b)) - (key(a) < key(b))
+    else:
+      if not compare:
+        self._comparator = lambda a, b: (a < b) - (a > b) if reverse \
+          else (a > b) - (a < b)
+      else:
+        self._comparator = lambda a, b: compare(b, a) if reverse \
+          else compare(a, b)
+      key = lambda elem: cmp_to_key(self._comparator)(elem)
+
+    self._compare = compare
+    self._num_quantiles = num_quantiles
+    self._buffer_size = buffer_size
+    self._num_buffers = num_buffers
+    self._key = key
+
+  @classmethod
+  def create(cls, num_quantiles, epsilon=None, max_num_elements=None,
```

Comment:
We do not seem to be exposing epsilon and max_num_elements from the transform. How often these parameters would be used? Is defaults good enough for most cases?

Reply:
Default values are good enough for most of the cases. But there might be the cases where a user wants to customize the maximum input size and epsilon.

Reply:
Sounds good.